File: 2- Regression/Polinomial regression/python/polynomial_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 27 13:50:26 2017

Polynomial regression
Use both a linear and a poliinomial regression to show the different results


@author: Ilaria
"""

# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

#we don't apply training and test set because we don't have enough data
# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)
"""

#fitting linear regression
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)

#fitting polinomial regression 
from sklearn.preprocessing import PolynomialFeatures
poly_reg =  PolynomialFeatures(degree = 4)
X_poli = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poli, y)

#visualize linear regression results
plt.scatter(X, y, color= 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Reality vs Bluff(linear regression)')
plt.xlabel('Position label')
plt.ylabel('Salary')
plt.show()


#visualize polinomial regression results
plt.scatter(X, y, color= 'red')
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color = 'blue')
plt.title('Reality vs Bluff(polynomial regression)')
plt.xlabel('Position label')
plt.ylabel('Salary')
plt.show()


#Predicting a new result with Linear Regression
lin_reg.predict(6.5)


#Predicting a new result with Polinomial regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: 2- Regression/Polinomial regression/R/polynomial_regression.R
#POLYNOMIAL REGRESSION
#import data 
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]


#let's split between train and test set
#library(caTools)
#set.seed(123)
##true goes to training, false to test
#split= sample.split(dataset$Profit, SplitRatio = 0.8)
#training_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

#fitting linear regression
lin_reg = lm(formula = Salary ~., data = dataset)
summary(lin_reg)

#fitting polinomial regression
#add poly level to the dataset
dataset$Level2= dataset$Level^2 
poly_reg = lm(formula = Salary ~., data = dataset)
summary(poly_reg)

#visualize linear regression
library(ggplot2)
ggplot()+ 
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)), colour = 'blue') + 
  ggtitle('Truth vs bluff linear regression') + 
  xlab ('level') + 
  ylab('salary')


#visualize polynomial regression 
library(ggplot2)
ggplot()+ 
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'blue') + 
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)), colour = 'blue') + 
  ggtitle('Truth vs bluff polynomial regression') + 
  xlab ('level') + 
  ylab('salary')

#predict a new result with linear regression
y_pred = predict(lin_reg, data.frame(Level=6.5))

#predict a new result with poly regression
y_pred = predict(lin_reg, data.frame(Level=6.5,
                                     Level2=6.5^2,
                                     Level3=6.5^3,
                                     Level4=6.5^4))

